### PR TITLE
Initial support for runevm contract

### PR DESCRIPTION
**Pull request by s1na**
This PR adds runevm as a system contract. I wanted to be able to test runevm via `testeth`, and made these changes to facilitate that. Not sure if this is something that you'd want merged.

Just now when I wanted to push my branch name conflicted with https://github.com/ewasm/hera/tree/runevm and I found this commit https://github.com/ewasm/hera/commit/866e12c90f644d834eb68836c700ba82c3d90267 (Update: just saw the PR https://github.com/ewasm/hera/pull/375) which would have saved me a day! :)

Three points about the PR:

- I couldn't get `context->host->call` to work, and directly use the `WasmEngine` to call runevm.
- When the runevm instance was being instantiated it would cause out-of-bounds memory access so I added the `module.memory.initial = 20` to avoid that. Not sure how to properly handle this.
- Added a flag `disable-interface-metering`, as runevm currently measures all gas costs itself (again, not sure if this is good or not).

### CODE REVIEW

File: src/hera.cpp [hunk 1]
```diff
@@ -73,6 +74,7 @@ const map<string, hera_evm1mode> evm1mode_options {
   { "reject", hera_evm1mode::reject },
   { "fallback", hera_evm1mode::fallback },
   { "evm2wasm", hera_evm1mode::evm2wasm_contract },
+  { "runevm", hera_evm1mode::runevm_contract },
```
Comment (axic) [inline src/hera.cpp]: Shouldn't need a new entry, it follows the same external API as evm2wasm. Just use it as "evm2wasm".

File: src/hera.cpp [hunk 2]
```diff
@@ -215,7 +262,7 @@ evmc_result hera_execute(
     heraAssert(rev == EVMC_BYZANTIUM, "Only Byzantium supported.");
     heraAssert(msg->gas >= 0, "EVMC supplied negative startgas");
 
-    bool meterInterfaceGas = true;
+    bool meterInterfaceGas = hera->meter_interface;
```
Comment (axic) [inline src/hera.cpp]: Instead of this, just uncomment these lines:
```
      case hera_evm1mode::evm2wasm_contract:
        run_code = evm2wasm(context, run_code);
        ensureCondition(run_code.size() > 8, ContractValidationFailure, "Transcompiling via evm2wasm failed");
        // TODO: enable this once evm2wasm does metering of interfaces
        // meterInterfaceGas = false;
break;
```

File: src/binaryen.cpp [hunk 3]
```diff
@@ -492,6 +492,7 @@ ExecutionResult BinaryenEngine::execute(
 
   // NOTE: DO NOT use the optimiser here, it will conflict with metering
 
+  module.memory.initial = 20;
```
Comment (axic) [inline src/binaryen.cpp]: Why is this needed?

File: src/hera.cpp (hunk 1 above)
Comment (s1na) [reply src/hera.cpp]: Can't get the `evm2wasm` mode to work. `callSystemContract` returns empty, as the address has no code associated with it (https://github.com/ethereum/aleth/blob/master/libethereum/Executive.cpp#L342). I expected aleth's evmc to redirect the call to hera again. Here's the command I use:

```
./testeth -t GeneralStateTests/stExample -- --singlenet Byzantium --singletest "add11" --vm ~/dev/ewasm/hera/build/src/libhera.so --evmc evm1mode=evm2wasm --evmc sys:evm2wasm=runevm2.wasm --verbosity 2
```

File: src/binaryen.cpp (hunk 3 above)
Comment (s1na) [reply src/binaryen.cpp]: Runevm's deployer ran out of memory. This was a temporary work-around. Should chisel's deployer maybe run `grow_memory`?

Comment (axic) [reply src/binaryen.cpp]: Right, it is due to this: https://github.com/wasmx/wasm-chisel/issues/53

File: src/hera.cpp (hunk 1 above)
Comment (axic) [reply src/hera.cpp]: Ah you are correct. We had this issue for it: https://github.com/ethereum/evmc/issues/53

I think we may be better just merging your `runevm` function, but later we could generalise it to have "run locally" vs. "run through the state".

File: src/binaryen.cpp (hunk 3 above)
Comment (s1na) [reply src/binaryen.cpp]: Removed this line.

File: src/hera.cpp (hunk 2 above)
Comment (s1na) [reply src/hera.cpp]: Would it be okay to leave this as an evmc flag? it'd be easier to both with interface metering enabled/disabled

File: src/hera.cpp [hunk 4]
```diff
@@ -191,6 +195,49 @@ vector<uint8_t> evm2wasm(evmc_context* context, vector<uint8_t> const& input) {
   return ret;
 }
 
+// Calls the runevm contract.
+// @returns a wasm-based evm interpreter.
+vector<uint8_t> runevm(evmc_context* context, WasmEngine& engine, vector<uint8_t> code) {
```
Comment (axic) [inline src/hera.cpp]: Can you make a helper `locallyExecuteSystemContract` similar to `callSystemContract`?

File: src/hera.cpp [hunk 5]
```diff
@@ -270,9 +323,6 @@ evmc_result hera_execute(
       );
     }
 
-    heraAssert(hera->engine, "Wasm engine not set.");
-    WasmEngine& engine = *hera->engine;
```
Comment (axic) [inline src/hera.cpp]: Even though it has a terrible overhead, I think `locallyExecuteSystemContract` should create its own engine instance to avoid any memory/state leaking.

**Comment by s1na** [comment]
Updated, can you have a look @axic?

File: src/hera.cpp [hunk 6]
```diff
@@ -67,34 +68,39 @@ const map<string, WasmEngineCreateFn> wasm_engine_map {
...
-  { "evm2wasm", hera_evm1mode::evm2wasm_contract },
-};
-
-struct hera_instance : evmc_instance {
-  unique_ptr<WasmEngine> engine{
+WasmEngineCreateFn wasmEngineCreateFn =
 // This is the order of preference.
 #if HERA_BINARYEN
-    new BinaryenEngine
+    BinaryenEngine::create
```
Comment (axic) [inline src/hera.cpp]: @chfast hmm, why did we had the static `create()` if we only ever used the default constructor?

File: src/hera.cpp [hunk 7]
```diff
@@ -247,6 +319,9 @@ evmc_result hera_execute(
         HERA_DEBUG << "Non-WebAssembly input, failure.\n";
         ret.status_code = EVMC_FAILURE;
         return ret;
+      case hera_evm1mode::runevm_contract:
+        run_code = runevm(context, hera->contract_preload_list[runevmAddress]);
+        ensureCondition(run_code.size() > 8, ContractValidationFailure, "Interpreting via runevm failed");
```
Comment (axic) [inline src/hera.cpp]: Need to set `meterInterfaceGas = false` here.

File: src/hera.cpp [hunk 8]
```diff
@@ -422,10 +497,19 @@ evmc_set_option_result hera_set_option(
     return EVMC_SET_OPTION_INVALID_VALUE;
   }
 
+  if (strcmp(name, "disable-interface-metering") == 0) {
```
Comment (axic) [inline src/hera.cpp]: I think this isn't needed anymore.

File: src/hera.cpp [hunk 9]
```diff
@@ -132,6 +138,40 @@ pair<evmc_status_code, bytes> callSystemContract(
...
+
+  unique_ptr<WasmEngine> engine = wasmEngineCreateFn();
+  ExecutionResult result = engine->execute(context, code, state_code, message, false);
+
+  bytes ret;
+  evmc_status_code status = result.isRevert ? EVMC_REVERT : EVMC_SUCCESS;
+  if (status == EVMC_SUCCESS && result.returnValue.size() > 0)
+    ret = move(result.returnValue);
+
+  heraAssert(result.gasLeft > 0, "Runevm ran out of gas.");
```
Comment (axic) [inline src/hera.cpp]: I don't think we need this assertion. The `runevm()` helper below will ensure this.

File: src/hera.cpp [hunk 10]
```diff
@@ -189,6 +229,38 @@ bytes evm2wasm(evmc_context* context, bytes_view input) {
...
+      gas,
+      {},
+      code,
+      code
+  );
+
+  HERA_DEBUG << "runevm done (output " << ret.size() << " bytes) with status=" << status << "\n";
+
+  heraAssert(ret.size() > 0, "Runevm returned empty.");
+  heraAssert(hasWasmPreamble(ret), "Runevm result has no wasm preamble.");
```
Comment (axic) [inline src/hera.cpp]: Not entirely sure, but I think we should use `ensureCondition` here in both cases.

File: src/hera.cpp [hunk 11]
```diff
@@ -425,7 +505,8 @@ evmc_set_option_result hera_set_option(
   if (strcmp(name, "engine") == 0) {
     auto it = wasm_engine_map.find(value);
     if (it != wasm_engine_map.end()) {
-      hera->engine = it->second();
+      wasmEngineCreateFn = it->second;
+      hera->engine = wasmEngineCreateFn();
```
Comment (axic) [inline src/hera.cpp]: Sorry, you were correct, `.execute()` is stateless at the moment. Would you mind if I remove the changes regarding the engines and merge this PR?

And then could you create a new PR changing the handling of the engine, because that might be a bug currently, but want to have @chfast's opinion.

Comment (s1na) [reply src/hera.cpp]: Sure, feel free to modify the PR.

File: src/hera.cpp (hunk 6 above)
Comment (chfast) [reply src/hera.cpp]: I don't remember anything about it. If the default constructor works for this case it's good choice.

**Comment by axic** [comment]
@chfast can you please review this PR?

File: src/hera.cpp [hunk 12]
```diff
@@ -132,6 +137,38 @@ pair<evmc_status_code, bytes> callSystemContract(
...
+    .value = {},
+    .create2_salt = {},
+  };
+
+  unique_ptr<WasmEngine> engine = wasmEngineCreateFn();
+  ExecutionResult result = engine->execute(context, code, state_code, message, false);
+
+  bytes ret;
+  evmc_status_code status = result.isRevert ? EVMC_REVERT : EVMC_SUCCESS;
+  if (status == EVMC_SUCCESS && result.returnValue.size() > 0)
```
Comment (chfast) [inline src/hera.cpp]:
```suggestion
  if (status == EVMC_SUCCESS && !result.returnValue.empty())
```
but the `returnValue` is not really needed.

File: src/hera.cpp [hunk 13]
```diff
@@ -132,6 +137,38 @@ pair<evmc_status_code, bytes> callSystemContract(
...
+    ret = move(result.returnValue);
+
+  return {status, ret};
```
Comment (chfast) [inline src/hera.cpp]: You have to move `ret` again.
